api/client/client.go

```diff
@@ -54,48 +52,116 @@ func init() {
+	dialer ContextDialer
+
 	// grpc is the gRPC client specification for the auth server.
 	grpc proto.AuthServiceClient
+
 	// conn is a grpc connection to the auth server.
 	conn *grpc.ClientConn
+
 	// closedFlag is set to indicate that the services are closed.
 	closedFlag int32
 }
 
-// NewClient returns a new auth client that uses mutual TLS authentication and
-// connects to the remote server using the Dialer or Addrs given in Config.
-func NewClient(cfg Config) (*Client, error) {
+// New creates a new API client with a connection to a Teleport server. The connection is
+// formed using the Dialer and CredentialsProviders given in config. If Addrs are provided, a
+// basic Dialer will be created. If Credentials are provided, a basic CredentialsProvider will
+// will be added to the list of CredentialsProviders.
+func New(cfg Config) (*Client, error) {
 	if err := cfg.CheckAndSetDefaults(); err != nil {
 		return nil, trace.Wrap(err)
 	}
 
 	c := &Client{c: cfg}
-	dialer := grpc.WithContextDialer(func(ctx context.Context, addr string) (net.Conn, error) {
-		if c.isClosed() {
-			return nil, trace.ConnectionProblem(nil, "client is closed")
+	if err := c.connect(cfg.NoPingCheck); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return c, nil
+}
+
+func (c *Client) getDialer(creds Credentials) (ContextDialer, error) {
+	dialer, err := creds.Dialer()
+	if err != nil {
+		if c.c.Dialer != nil {
+			return c.c.Dialer, nil
+		}
+		if len(c.c.Addrs) == 0 {
+			return nil, trace.BadParameter("no dialer in credentials, configuration, or addresses provided")
 		}
-		conn, err := c.c.Dialer.DialContext(ctx, "tcp", addr)
+		dialer, err = NewAddrDialer(c.c.Addrs, c.c.KeepAlivePeriod, c.c.DialTimeout)
 		if err != nil {
-			return nil, trace.ConnectionProblem(err, "failed to dial")
+			return nil, trace.Wrap(err)
 		}
-		return conn, nil
-	})
+	}
+	return dialer, nil
+}
 
+func (c *Client) connect(noPingCheck bool) error {
+	// Loop over credentials and use first successfully one.
 	var err error
-	if c.conn, err = grpc.Dial(constants.APIDomain,
-		dialer,
-		grpc.WithTransportCredentials(credentials.NewTLS(c.c.TLS)),
-		grpc.WithKeepaliveParams(keepalive.ClientParameters{
-			Time:                c.c.KeepAlivePeriod,
-			Timeout:             c.c.KeepAlivePeriod * time.Duration(c.c.KeepAliveCount),
-			PermitWithoutStream: true,
-		}),
-	); err != nil {
-		return nil, trail.FromGRPC(err)
+	var errs []error
+	for _, creds := range c.c.Credentials {
+		// Load *tls.Config from the provided credentials.
+		c.tlsConfig, err = creds.Config()
+		if err != nil {
+			errs = append(errs, trace.Wrap(err))
+			continue
+		}
+
+		// Build a dialer, prefer a dialer from credentials. If no fallback to the
+		// passed in dialer and then list of addresses.
+		c.dialer, err = c.getDialer(creds)
+		if err != nil {
+			errs = append(errs, trace.Wrap(err))
+			continue
+		}
+		grpcDialer := grpc.WithContextDialer(func(ctx context.Context, addr string) (net.Conn, error) {
```

Comment:
Nit: Why not hide this behind `getGRPCDialer`? Or even just make `getDialer` return grpc dialer, I don't think it's used for anything else?